README.md

# progress_bars

  

Circular progress bars with interesting animation movements.
  

## Getting Started

  To use this plugin, add _progress_bars_ as a dependency in your pubspec.yaml file.

## Circle Progress Bar
<img src="https://github.com/swift0vinay/progress_bars/blob/master/demo/demo.gif" width="250" height="500"/>


Circle Progress Bar has 5 properties :

 1. _progressColor_ : Represents the Color of the arc progressing .
 2. _backgroundColor_ : Represents the background of arc progressing .
 3.  _child_ : Represents the widget appearing in the center of progress bar .
 4. _size_ : Represents the radius of the progress bar .
 5. _arcWidth_ : Represents the sectorLength of arc .


Example :

````
import 'package:flutter/material.dart';
import 'package:progress_bars/progress_bars.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Page(),
      ),
    );
  }
}

class Page extends StatefulWidget {
  @override
  _PageState createState() => _PageState();
}

class _PageState extends State<Page> {
  String text = 'Press To Load';
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Align(
      alignment: Alignment(0, 0.6),
      child: FlatButton(
        child: Text(
          text,
          style: TextStyle(
            color: Colors.white,
            fontSize: 15.0,
            fontWeight: FontWeight.bold,
          ),
        ),
        color: Colors.orange,
        onPressed: () {
          showDialog(
              context: context,
              barrierDismissible: false,
              builder: (context) {
                return Center(
                  child: CircleProgressBar(
                    backgroundColor: Colors.black54,
                    progressColor: Colors.orange,
                    size: 40,
                    child: SizedBox(
                      child: FlutterLogo(),
                    ),
                  ),
                );
              });
          Future.delayed(Duration(seconds: 10)).then(
            (value) {
              Navigator.pop(context);
              setState(
                () {
                  text = "LOADED";
                },
              );
            },
          );
        },
      ),
    ));
  }
}
````

CHANGELOG.md
## [0.0.3] - Upgraded Animations

* Animations are made more eye-catching .

## [0.0.2] - Bug Fixing

* Bug Fixing .

## [0.0.1] - First Release

* CircleProgressBar added .

test/progress_bars_test.dart
import 'package:flutter/material.dart';
import 'package:progress_bars/progress_bars.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Page(),
      ),
    );
  }
}

class Page extends StatefulWidget {
  @override
  _PageState createState() => _PageState();
}

class _PageState extends State<Page> {
  String text = 'Press To Load';
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Align(
      alignment: Alignment(0, 0.6),
      child: FlatButton(
        child: Text(
          text,
          style: TextStyle(
            color: Colors.white,
            fontSize: 15.0,
            fontWeight: FontWeight.bold,
          ),
        ),
        color: Colors.orange,
        onPressed: () {
          showDialog(
              context: context,
              barrierDismissible: false,
              builder: (context) {
                return Center(
                  child: CircleProgressBar(
                    backgroundColor: Colors.black54,
                    progressColor: Colors.orange,
                    size: 40,
                    child: SizedBox(
                      child: FlutterLogo(),
                    ),
                  ),
                );
              });
          Future.delayed(Duration(seconds: 10)).then(
            (value) {
              Navigator.pop(context);
              setState(
                () {
                  text = "LOADED";
                },
              );
            },
          );
        },
      ),
    ));
  }
}


lib/circle_progress_bar/circle_progress_painter.dart
import 'dart:math';
import 'package:flutter/material.dart';

class CircleProgressPainter extends CustomPainter {
  /// [backgroundColor] represents the background of arc progressing .
  /// @Default is [Colors.grey[200]]
  final backgroundColor;

  /// [progressColor] represents the Color of the arc progressing .
  /// @Default is [Colors.green[600]]
  final progressColor;

  /// [height] [width] are dimension of canvas ,
  /// [rightProgress] [leftProgress] are leftAnimation and rightAnimation values respectively ,
  double height, width, rightProgress, leftProgress, combinedProgress, arcWidth;
  bool combined;
  String version;
  CircleProgressPainter({
    this.version,
    this.backgroundColor,
    this.leftProgress,
    this.rightProgress,
    this.progressColor,
    this.height,
    this.arcWidth,
    this.width,
    this.combinedProgress,
    this.combined,
  });

  /// [degToRad] function to convert angle in degrees to radians .
  num degToRad(num deg) => deg * (pi / 180.0);

  @override
  void paint(Canvas canvas, Size size) {
    /// [radius] is the radius of progress bar .
    double radius = height * 0.5;

    Paint pr = Paint()
      ..color = progressColor ?? Colors.green[600]
      ..style = PaintingStyle.stroke
      ..strokeWidth = radius * 0.25
      ..strokeCap = StrokeCap.round;

    Paint paint = Paint()
      ..color = backgroundColor ?? Colors.grey[200]
      ..style = PaintingStyle.stroke
      ..strokeWidth = radius * 0.25
      ..strokeCap = StrokeCap.round;

    /// [center] is the center of progress bar .
    Offset center = Offset(width / 2, height / 2);

    canvas.drawCircle(center, radius, paint);

    Rect rect = Rect.fromCircle(center: center, radius: radius);

    if (combined) {
      drawArc(
          canvas, rect, degToRad(combinedProgress), degToRad(-arcWidth), pr);
    } else {
      if (version == 'F') {
        drawArc(canvas, rect, degToRad(270 - arcWidth / 3),
            degToRad(2 * arcWidth / 3), pr);
        drawArc(
            canvas, rect, degToRad(leftProgress), degToRad(-arcWidth / 6), pr);
        drawArc(
            canvas, rect, degToRad(rightProgress), degToRad(arcWidth / 6), pr);
      } else if (version == 'S') {
        drawArc(canvas, rect, degToRad(90 + arcWidth / 6),
            degToRad(-arcWidth / 3), pr);
        drawArc(canvas, rect, degToRad(270 - arcWidth / 6),
            degToRad(arcWidth / 3), pr);
        drawArc(
            canvas, rect, degToRad(leftProgress), degToRad(-arcWidth / 6), pr);
        drawArc(
            canvas, rect, degToRad(rightProgress), degToRad(arcWidth / 6), pr);
      } else {
        drawArc(canvas, rect, degToRad(90 + arcWidth / 3),
            degToRad(-2 * arcWidth / 3), pr);
        drawArc(
            canvas, rect, degToRad(leftProgress), degToRad(-arcWidth / 6), pr);
        drawArc(
            canvas, rect, degToRad(rightProgress), degToRad(arcWidth / 6), pr);
      }
    }
  }

  /// [drawArc] uses the [canvas.drawArc] method only .
  drawArc(Canvas canvas, Rect rect, double startAngle, double sweepAngle,
      Paint pr) {
    canvas.drawArc(rect, startAngle, sweepAngle, false, pr);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}


lib/circle_progress_bar/circle_progress_bar.dart
import 'package:flutter/material.dart';
import 'circle_progress_painter.dart';

class CircleProgressBar extends StatefulWidget {
  /// [progressColor] represents the Color of the arc progressing .
  /// @Default is [Colors.green[600]]
  final Color progressColor;

  /// [backgroundColor] represents the background of arc progressing .
  /// @Default is [Colors.grey[200]]
  final Color backgroundColor;

  /// [child] represents the widget appearing in the center of progress bar .
  final Widget child;

  /// [size] represents the radius of the progress bar
  /// @Default is 40
  final double size;

  /// [arcWidth] represents the sectorLength of arc
  /// @Default is 50
  final double arcWidth;

  CircleProgressBar({
    this.progressColor,
    this.backgroundColor,
    this.child,
    this.size: 40,
    this.arcWidth: 50,
  });

  @override
  _CircleProgressBar createState() => _CircleProgressBar();
}

class _CircleProgressBar extends State<CircleProgressBar>
    with TickerProviderStateMixin {
  /// [leftAnimation] defines arc movement towards left to bottom .
  Animation<double> leftAnimation;

  /// [rightAnimation] defines arc movement towards right to bottom .
  Animation<double> rightAnimation;

  /// [combinedAnimation] defines combined arc movements in various directions .
  Animation<double> combinedAnimation;

  /// [lrController] controller which controls [leftAnimation] & [rightAnimation] .
  AnimationController lrController;

  /// [combinedController] controller for [combinedController] .
  AnimationController combinedController;

  double height, width, leftProgress, rightProgress, combinedProgress;
  bool combined = true;
  String version = "F";
  @override
  void initState() {
    super.initState();
    height = width = 2 * widget.size;
    super.initState();
    startSequence();
  }

  @override
  void dispose() {
    lrController.dispose();
    combinedController.dispose();
    super.dispose();
  }

  startSequence() {
    lrController = AnimationController(
        vsync: this, duration: Duration(milliseconds: 1000));
    combinedController =
        AnimationController(vsync: this, duration: Duration(seconds: 4));

    createSequence();

    leftAnimation
        .addListener(() => setState(() => leftProgress = leftAnimation.value));
    rightAnimation.addListener(
        () => setState(() => rightProgress = rightAnimation.value));
    combinedAnimation.addListener(
        () => setState(() => combinedProgress = combinedAnimation.value));

    lrController.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        if (version == 'F') {
          setState(() {
            version = 'S';
            leftAnimation = Tween<double>(
                    begin: 270 - this.widget.arcWidth / 6,
                    end: 90 + this.widget.arcWidth / 3)
                .animate(lrController);
            rightAnimation = Tween<double>(
                    begin: 270 + this.widget.arcWidth / 6,
                    end: 450 - this.widget.arcWidth / 3)
                .animate(lrController);
          });
          lrController.forward(from: 0.0);
        } else if (version == 'S') {
          setState(() {
            version = 'T';
            leftAnimation =
                Tween<double>(begin: 270, end: 90 + this.widget.arcWidth / 2)
                    .animate(lrController);
            rightAnimation =
                Tween<double>(begin: 270, end: 450 - this.widget.arcWidth / 2)
                    .animate(lrController);
          });
          lrController.forward(from: 0.0);
        } else {
          lrController.stop();
        }
      }
    });

    combinedController.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        combined = false;
        leftAnimation = Tween<double>(
                begin: 270 - this.widget.arcWidth / 3,
                end: 90 + this.widget.arcWidth / 6)
            .animate(lrController);
        rightAnimation = Tween<double>(
                begin: 270 + this.widget.arcWidth / 3,
                end: 450 - this.widget.arcWidth / 6)
            .animate(lrController);
        lrController.forward();
        Future.delayed(Duration(milliseconds: 3000), () {
          version = 'F';
          combined = true;
          lrController.reset();
          combinedController.forward(from: 0.0);
        });
      }
    });

    combinedController.forward();
  }

  /// [createSequence] is an helper function to initialize the animation variable .
  createSequence() {
    leftAnimation = Tween<double>(begin: 270, end: (90 + widget.arcWidth / 2))
        .animate(lrController);

    leftProgress = leftAnimation.value;

    rightAnimation = Tween<double>(begin: 270, end: (450 - widget.arcWidth / 2))
        .animate(lrController);

    rightProgress = leftAnimation.value;

    combinedAnimation = TweenSequence([
      TweenSequenceItem(
        tween: Tween<double>(
            begin: (90 + widget.arcWidth / 2),
            end: (180 + widget.arcWidth / 2)),
        weight: 5,
      ),
      TweenSequenceItem(
        tween: Tween<double>(
            begin: (180 + widget.arcWidth / 2), end: (widget.arcWidth / 2)),
        weight: 10,
      ),
      TweenSequenceItem(
        tween: Tween<double>(
            begin: (widget.arcWidth / 2), end: (270 + widget.arcWidth / 2)),
        weight: 20,
      ),
    ]).animate(combinedController);
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: CustomPaint(
        foregroundPainter: CircleProgressPainter(
          backgroundColor: this.widget.backgroundColor,
          progressColor: this.widget.progressColor,
          height: height,
          width: width,
          leftProgress: leftProgress,
          rightProgress: rightProgress,
          combined: combined,
          combinedProgress: combinedProgress,
          arcWidth: widget.arcWidth,
          version: version,
        ),
        child: Container(
          height: height,
          width: width,
          child: Container(
              height: height / 2,
              width: width / 2,
              child: Center(child: widget.child)),
        ),
      ),
    );
  }
}
